engine/src/main/java/org/hibernate/validator/internal/properties/javabean/JavaBean.java

```diff
@@ -6,32 +6,210 @@
  */
 package org.hibernate.validator.internal.properties.javabean;
 
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Field;
+import java.lang.reflect.Member;
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import java.lang.reflect.Type;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+import java.util.ArrayList;
 import java.util.Arrays;
-import java.util.stream.Stream;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
 
-import org.hibernate.validator.internal.properties.Property;
 import org.hibernate.validator.internal.properties.ConstrainableType;
-import org.hibernate.validator.internal.util.ReflectionHelper;
+import org.hibernate.validator.internal.util.CollectionHelper;
+import org.hibernate.validator.internal.util.privilegedactions.GetDeclaredConstructors;
+import org.hibernate.validator.internal.util.privilegedactions.GetDeclaredFields;
+import org.hibernate.validator.internal.util.privilegedactions.GetDeclaredMethods;
+import org.hibernate.validator.internal.util.privilegedactions.GetMethods;
+import org.hibernate.validator.spi.properties.ConstrainableExecutable;
+import org.hibernate.validator.spi.properties.GetterPropertySelectionStrategy;
 
 /**
+ * {@link ConstrainableType} that describes a JavaBean.
+ *
  * @author Marko Bekhta
  */
 public class JavaBean implements ConstrainableType {
 
-	private final Class<?> clazz;
+	private final String name;
+	private final List<JavaBeanField> beanFields;
+	private final List<JavaBeanMethod> beanMethods;
+	private final List<JavaBeanGetter> beanGetters;
+	private final List<JavaBeanConstructor> beanConstructors;
```

Comment:
So... this is the big issue I have with this patch. At one point, you were talking about using `ConstrainableType` as a vehicle for the class in various parts of the code (with the idea to replace `getDeclaringClass()`). From what I see of the `getDeclaringClass()` usage, we would have to keep it after the bootstrapping of the metadata. And I'm not sure I want to keep all these references. Keep in mind that we're not only talking about the constrained elements here but all the elements of a bean.

So I have a couple of comments:
- I think these elements should not be located here but rather in an object that has a limited lifespan limited in scope to the metadata discovery;
- The usage pattern for the XML and the programmatic API is very suboptimal. Now that I think about it, I have probably the answer of why the strategy is so simplified now, it's due to the removal of the reflection calls.

I think I would completely separate it from the `properties` hierarchy (I think we will need a `JavaBean` there but a simpler one). Maybe you can call this one `BeanMetaDataDiscoverer`? I wouldn't implement any interface as it's just something we will want to use as is.

I also think that we should use ``Map``s instead of ``List``s to avoid iterating again and again on the elements. Is there a reason why you didn't use ``Map``s?